lightning/src/util/events.rs

```diff
@@ -380,9 +381,32 @@ pub trait MessageSendEventsProvider {
 	fn get_and_clear_pending_msg_events(&self) -> Vec<MessageSendEvent>;
 }
 
-/// A trait indicating an object may generate events
+/// A trait indicating an object may generate events.
+///
+/// Events are processed by a handler given to [`process_pending_events`]. Therefore, implementors
+/// should be mindful to avoid lock reentrancy if a handler may call back into the provider.
```

Comment:
Should we make it a part of the contract that callbacks into the provider are allowed/don't run locked, except for (possibly) serialization (in the case of a ChannelManager)? Its kinda hard to work with this API without that guarantee, but its kinda annoying to specify it.

Reply:
For `ChannelManager`, both `process_pending_events` and `handle_event` take a read lock. The write lock is only taken after notification (i.e., after `process_pending_events` has returned) -- the handler itself doesn't serialize `ChannelManager`.

So I think the general advice is "don't introduce a deadlock". But now that I think of it, does `RwLock` allow reentrant reads? It appears so though the docs say it _might_ panic.

https://doc.rust-lang.org/std/sync/struct.RwLock.html#panics

Reply:
> So I think the general advice is "don't introduce a deadlock".

Right, but the docs don't *really* give you enough information to know what would and what would not cause a deadlock. Reading it naively, I'd think I need to queue all the events and then call back to ChannelManager separately as calling back to it may cause a deadlock. Providing more information would let users do more without fear, though we have to be precise and decide what we're willing to commit to in the API and what not.

> does RwLock allow reentrant reads?

I assumed it would...I mean you're allowed to take as many read locks as you want....

Reply:
> Right, but the docs don't _really_ give you enough information to know what would and what would not cause a deadlock. Reading it naively, I'd think I need to queue all the events and then call back to ChannelManager separately as calling back to it may cause a deadlock. Providing more information would let users do more without fear, though we have to be precise and decide what we're willing to commit to in the API and what not.

Understood, but these aren't `ChannelManager`'s docs. And note that this line is talking about someone implementing `EventsProvider` not `EventHandler`.  It's really up to an implementor of `EventsProvider` to determine what is safe for a handler to call. We don't need to restrict it at the trait level.

Reply:
Right, I guess then the docs are somewhat confusing to me - I read them as "user, you should be careful of reentrancy when calling methods here", not "note to RL implementors, be careful of reentrancy". We should have some docs per-implementor that talks about reentrancy, I guess, though sadly Rust really hides the impl-level docs for a trait implementation, making them hard to find :(.

Reply:
Re-wrote the docs to separate implementation and uses. Also added docs to `ChannelManager` and `ChainMonitor` implementations. Let me know if anything isn't clear or could be expanded upon.

Reply:
Having a whole section dedicated to `Implementations` that contains instructions for how to implement just feels really wrong in public docs here? In practice, I can't imagine we ever expect users to implement this, so it becomes just a list of instructions to ourselves in public docs?

Reply:
Combined and rephrased in terms of requirements.